jepsen/src/jepsen/control/k8s.clj

```diff
@@ -0,0 +1,111 @@
+(ns jepsen.control.k8s
+  "The recommended way is to use SSH to setup and teardown databases.
+  It's however sometimes conveniet to be able to setup and teardown
+  the databases using `kubectl` instead, which is what this namespace
+  helps you do. Use at your own risk, this is an unsupported way
+  of running Jepsen."
+  (:require [clojure.java.shell :refer [sh]]
+            [slingshot.slingshot :refer [throw+]]
+            [jepsen.control :as c]
+            [clojure.string :refer [split-lines, trim]]
+            [clojure.tools.logging :refer [info]])
+  (:import (jepsen.control Remote)))
+
+(defn exec
+  "Execute a shell command on a pod."
+  [context namespace pod-name {:keys [cmd] :as opts}]
+  (apply sh
+         "kubectl"
+         "exec"
+         (c/escape pod-name)
+         context
+         namespace
+         "--"
+         "sh"
+         "-c"
+         cmd
```

Comment:
I don't remember sh's escaping semantics off the top of my head, and I don't know kubectl's at all, so I'd want to be really cautious here about weird inputs--it'd be worth testing with shell magic characters--quotes, semicolons, ampersands, that kind of thing.

Reply:
in my test, there is some complexity behing "--". Effectively, I do believe they have deprecated the previous way of exec-ing commands and are replacing it now with the full freedom exactly behind "--" for all kinds of special cases. Can you give an example of whatever kind? it's easy to run against any k8s context, so I can quickly verify. Thanks.

Reply:
Like I said, single and double quotes, semicolons, ampersands, dashes, pipes, redirection brackets, all that good stuff. You'll wanna make sure that gets correctly escaped, or this could end up exploding in some really awkward ways. Like, delete-your-home-directory kinda stuff, haha. :)

Reply:
unfortunately, I don't see how to formalize that in local tests without a significantly larger effort, but I'll do some in repl and report the results here for comparison.

Reply:
`foobar=> (k8s/exec "" "" "config-56ff747f6b-rqsjz" {:cmd "export FOOBAR=tmp && cd /`echo $FOOBAR` ; ls -la $(pwd)"})
{:exit 0, :out "total 8\ndrwxrwxrwt 1 root root 4096 Jul 17 19:24 .\ndrwxr-xr-x 1 root root 4096 Jul 16 15:49 ..\n", :err ""}
foobar=>`

Reply:
lol, I spent more time escaping here int the inline editor than running the actual test, but example above should get the point across: here, I'm running remote's exec function with all kinds of linux shell trickery / special characters, as if it would be called from controls/sh itself, against some random pod in my currently default k8s context/namespace. I believe on the way :cmd gets properly escaped already, the example anyway demostrates correctness IMO - we're working with an env FOOBAR to hold temp-folder, in which we change and from which we ls, all as expected.

Also, double-dashing sh -c after a kubectl exec is to my knowledge the safest way to avoid weird inputs and outputs.

Reply:
Cool cool. :)